packages/react-instantsearch-nextjs/src/InitializePromise.tsx

```diff
@@ -25,16 +25,30 @@ export function InitializePromise() {
   // later during hydration.
   let requestParamsList: SearchOptions[];
   search.mainHelper!.setClient({
+    ...search.mainHelper!.getClient(),
     search(queries) {
       requestParamsList = queries.map(({ params }) => params!);
       return search.client.search(queries);
     },
   });
 
+  resetWidgetId();
+
   const waitForResults = () =>
     new Promise<void>((resolve) => {
-      search.mainHelper!.derivedHelpers[0].on('result', () => {
-        resolve();
+      let searchReceived = false;
```

Comment:
Suggestion for a different more "promise-like" approach:

```js
Promise.all([
  search._hasSearchWidget && new Promise(
    (resolve) => search.mainHelper!.derivedHelpers[0].once('result', resolve)
  ),
  search._hasRecommendWidget && new Promise(
    (resolve) => search.mainHelper!.derivedHelpers[0].once('recommend:result', resolve)
  ),
].filter(Boolean)).then(resolve);
```

Reply:
I would have preferred it that way, unfortunately `_hasSearchWidget` and `_hasRecommendWidget` will always be `false` at this point as we initialize the promise before adding any widget.
This is one of the use cases where an event-based API makes sense. I could do that in the "legacy" `waitForResults` though, but maybe it's best to keep consistency.